### PR TITLE
sipify geometry classes

### PR DESCRIPTION
**Pull request by 3nids**
## Description
Include a few sentences describing the overall goals for this PR (pull request). If applicable also add screenshots.

## Checklist

> Reviewing is a process done by project maintainers, mostly on a volunteer basis. We try to keep the overhead as small as possible and appreciate if you help us to do so by completing the following items. Feel free to ask in a comment if you have troubles with any of them.

- [ ] Commit messages are descriptive and explain the rationale for changes
- [ ] Commits which fix bugs include `fixes #11111` in the commit message next to the description
- [ ] Commits which add new features are tagged with `[FEATURE]` in the commit message
- [ ] Commits which change the UI or existing user workflows are tagged with `[needs-docs]` in the commit message and containt sufficient information in the commit message to be documented
- [ ] I have read the [QGIS Coding Standards](https://docs.qgis.org/testing/en/docs/developers_guide/codingstandards.html) and this PR complies with them
- [ ] This PR passes all existing unit tests (test results will be reported by travis-ci after opening this PR)
- [ ] New unit tests have been added for core changes
- [ ] I have run [the `scripts/prepare-commit.sh` script](https://github.com/qgis/QGIS/blob/master/.github/CONTRIBUTE.md#contributing-to-qgis) before each commit

### CODE REVIEW

File: python/core/geometry/qgsabstractgeometry.sip [hunk 1]
```diff
@@ -1,406 +1,537 @@
...
+ :rtype: QgsRectangle
+%End
+
 
-    //mm-sql interface
-    /** Returns the inherent dimension of the geometry. For example, this is 0 for a point geometry,
-     * 1 for a linestring and 2 for a polygon.
-     */
     virtual int dimension() const = 0;
-    //virtual int coordDim() const { return mCoordDimension; }
```
Comment (nyalldawson) [inline python/core/geometry/qgsabstractgeometry.sip]: Why is this missing from sip?

File: python/core/geometry/qgsabstractgeometry.sip [hunk 2]
```diff
@@ -1,406 +1,537 @@
...
-     */
     bool isMeasure() const;
+%Docstring
+ Returns true if the geometry contains m values.
+ \see is3D
+ :rtype: bool
+%End
+
+
+    virtual QgsAbstractGeometry *boundary() const = 0;
```
Comment (nyalldawson) [inline python/core/geometry/qgsabstractgeometry.sip]: Factory

File: python/core/geometry/qgsabstractgeometry.sip [hunk 3]
```diff
@@ -1,406 +1,537 @@
...
 
-    /** Returns the point corresponding to a specified vertex id
-     */
     virtual QgsPointV2 vertexAt( QgsVertexId id ) const = 0;
+%Docstring
+ Returns the point corresponding to a specified vertex id
+ :rtype: QgsPointV2
+%End
+
+    virtual double closestSegment( const QgsPointV2 &pt, QgsPointV2 &segmentPt, QgsVertexId &vertexAfter, bool *leftOf, double epsilon ) const = 0;
```
Comment (nyalldawson) [inline python/core/geometry/qgsabstractgeometry.sip]: Should the arguments be /out/?

File: python/core/geometry/qgscircularstring.sip [hunk 4]
```diff
@@ -1,102 +1,121 @@
...
+class QgsCircularString: QgsCurve
 {
-%TypeHeaderCode
-#include <qgscircularstring.h>
+%Docstring
+ Circular string geometry type
+.. versionadded:: 2.10
+.. note::
+
+   this API is not considered stable and may change for 2.12
```
Comment (nyalldawson) [inline python/core/geometry/qgscircularstring.sip]: This line should be removed

File: python/core/geometry/qgscompoundcurve.sip [hunk 5]
```diff
@@ -1,94 +1,143 @@
...
+class QgsCompoundCurve: QgsCurve
 {
-%TypeHeaderCode
-#include <qgscompoundcurve.h>
+%Docstring
+ Compound curve geometry type
+.. versionadded:: 2.10
+.. note::
+
+   this API is not considered stable and may change for 2.12
```
Comment (nyalldawson) [inline python/core/geometry/qgscompoundcurve.sip]: Remove

File: python/core/geometry/qgscompoundcurve.sip [hunk 6]
```diff
@@ -1,94 +1,143 @@
...
+    const QgsCurve *curveAt( int i ) const;
+%Docstring
+ Returns the curve at the specified index.
+ :rtype: QgsCurve
+%End
 
-    /** Adds a curve to the geometr (takes ownership)
-     */
-    void addCurve( QgsCurve* c /Transfer/ );
+    void addCurve( QgsCurve *c ) /Transfer/;
```
Comment (nyalldawson) [inline python/core/geometry/qgscompoundcurve.sip]: This transfer is in the wrong place

File: python/core/geometry/qgscurvepolygon.sip [hunk 7]
```diff
@@ -1,108 +1,167 @@
...
+class QgsCurvePolygon: QgsSurface
 {
-%TypeHeaderCode
-#include <qgscurvepolygon.h>
+%Docstring
+ Curve polygon geometry type
+.. versionadded:: 2.10
+.. note::
+
+   this API is not considered stable and may change for 2.12
```
Comment (nyalldawson) [inline python/core/geometry/qgscurvepolygon.sip]: Remove

File: python/core/geometry/qgscurvepolygon.sip [hunk 8]
```diff
@@ -1,108 +1,167 @@
...
+    const QgsCurve *exteriorRing() const;
+%Docstring
+ :rtype: QgsCurve
+%End
+    const QgsCurve *interiorRing( int i ) const;
+%Docstring
+ :rtype: QgsCurve
+%End
+
+    virtual QgsPolygonV2 *toPolygon( double tolerance = M_PI_2 / 90, SegmentationToleranceType toleranceType = MaximumAngle ) const;
```
Comment (nyalldawson) [inline python/core/geometry/qgscurvepolygon.sip]: Factory

File: python/core/geometry/qgscurvepolygon.sip [hunk 9]
```diff
@@ -1,108 +1,167 @@
...
+%Docstring
+ Returns a new polygon geometry corresponding to a segmentized approximation
+ of the curve.
+ \param tolerance segmentation tolerance
+ \param toleranceType maximum segmentation angle or maximum difference between approximation and curve*
+ :rtype: QgsPolygonV2
+%End
+
+    virtual void setExteriorRing( QgsCurve *ring ) /Transfer/;
```
Comment (nyalldawson) [inline python/core/geometry/qgscurvepolygon.sip]: Transfer in wrong spot

File: python/core/geometry/qgscurvepolygon.sip [hunk 10]
```diff
@@ -1,108 +1,167 @@
...
+%Docstring
+ Sets the exterior ring of the polygon. The CurvePolygon type will be updated to match the dimensionality
+ of the exterior ring. For instance, setting a 2D exterior ring on a 3D CurvePolygon will drop the z dimension
+ from the CurvePolygon and all interior rings.
+ \param ring new exterior ring. Ownership is transferred to the CurvePolygon.
+ \see setInteriorRings()
+ \see exteriorRing()
+%End
+
+    void setInteriorRings( const QList<QgsCurve *> &rings ) /Transfer/;
```
Comment (nyalldawson) [inline python/core/geometry/qgscurvepolygon.sip]: Transfer misplaced

File: python/core/geometry/qgscurvepolygon.sip [hunk 11]
```diff
@@ -1,108 +1,167 @@
...
+%Docstring
+Sets all interior rings (takes ownership)
+%End
+    virtual void addInteriorRing( QgsCurve *ring );
```
Comment (nyalldawson) [inline python/core/geometry/qgscurvepolygon.sip]: Missing transfer

**Review by nyalldawson — COMMENTED** [review]
Just a partial review...

File: python/core/geometry/qgsgeometry.sip [hunk 12]
```diff
@@ -1,948 +1,1286 @@
...
+ :rtype: bool
+%End
+
+    bool moveVertex( const QgsPointV2 &p, int atVertex );
+%Docstring
+ Moves the vertex at the given position number
+  and item (first number is index 0)
+  to the given coordinates.
+//     *  Returns false if atVertex does not correspond to a valid vertex
+  on this geometry
```
Comment (nyalldawson) [inline python/core/geometry/qgsgeometry.sip]: looks like brackets in the comments throw out something

File: python/core/geometry/qgsgeometry.sip [hunk 13]
```diff
@@ -1,948 +1,1286 @@
...
+
+    QgsGeometry shortestLine( const QgsGeometry &other ) const;
+%Docstring
+ Returns the shortest line joining this geometry to another geometry.
+.. versionadded:: 2.14
+ \see nearestPoint()
+ :rtype: QgsGeometry
+%End
+
+    double closestVertexWithContext( const QgsPoint &point, int &atVertex ) const;
```
Comment (nyalldawson) [inline python/core/geometry/qgsgeometry.sip]: Missing /Out/

File: python/core/geometry/qgsgeometry.sip [hunk 14]
```diff
@@ -1,948 +1,1286 @@
...
+
+    int addRing( const QList<QgsPoint> &ring );
+%Docstring
+ Adds a new ring to this geometry. This makes only sense for polygon and multipolygons.
+:return: 0 in case of success (ring added), 1 problem with geometry type, 2 ring not closed,
+3 ring is not valid geometry, 4 ring not disjoint with existing rings, 5 no polygon found which contained the ring*
+ :rtype: int
+%End
+
+    int addRing( QgsCurve *ring );
```
Comment (nyalldawson) [inline python/core/geometry/qgsgeometry.sip]: /Transfer/

File: python/core/geometry/qgsgeometry.sip [hunk 15]
```diff
@@ -1,948 +1,1286 @@
...
+ Adds a new part to this geometry.
+ \param part part to add (ownership is transferred)
+ \param geomType default geometry type to create if no existing geometry
+ :return: 0 in case of success, 1 if not a multipolygon, 2 if ring is not a valid geometry, 3 if new polygon ring
+ not disjoint with existing polygons of the feature
+ :rtype: int
+%End
+
+
+    int addPart( const QgsGeometry &newPart );
```
Comment (nyalldawson) [inline python/core/geometry/qgsgeometry.sip]: Missing  /PyName=addPartGeometry/

File: python/core/geometry/qgsgeometry.sip [hunk 16]
```diff
@@ -1,948 +1,1286 @@
...
+
+    int addRing( QgsCurve *ring );
+%Docstring
+ Adds a new ring to this geometry. This makes only sense for polygon and multipolygons.
+:return: 0 in case of success (ring added), 1 problem with geometry type, 2 ring not closed,
+3 ring is not valid geometry, 4 ring not disjoint with existing rings, 5 no polygon found which contained the ring*
+ :rtype: int
+%End
+
+    int addPart( const QList<QgsPoint> &points, QgsWkbTypes::GeometryType geomType = QgsWkbTypes::UnknownGeometry );
```
Comment (nyalldawson) [inline python/core/geometry/qgsgeometry.sip]: Missing /PyName=addPoints/

File: python/core/geometry/qgsgeometry.sip [hunk 17]
```diff
@@ -1,948 +1,1286 @@
...
+%Docstring
+ Adds a new part to a the geometry.
+ \param points points describing part to add
+ \param geomType default geometry type to create if no existing geometry
+ :return: 0 in case of success, 1 if not a multipolygon, 2 if ring is not a valid geometry, 3 if new polygon ring
+ not disjoint with existing polygons of the feature
+ :rtype: int
+%End
+
+    int addPart( const QgsPointSequence &points, QgsWkbTypes::GeometryType geomType = QgsWkbTypes::UnknownGeometry );
```
Comment (nyalldawson) [inline python/core/geometry/qgsgeometry.sip]: Missing /PyName=addPointsV2/

File: python/core/geometry/qgsgeometrycollection.sip [hunk 18]
```diff
@@ -1,97 +1,144 @@
...
+class QgsGeometryCollection: QgsAbstractGeometry
 {
-%TypeHeaderCode
-#include <qgsgeometrycollection.h>
+%Docstring
+ Geometry collection
+.. versionadded:: 2.10
+.. note::
+
+   this API is not considered stable and may change for 2.12
```
Comment (nyalldawson) [inline python/core/geometry/qgsgeometrycollection.sip]: remove

File: python/core/geometry/qgsgeometrycollection.sip [hunk 19]
```diff
@@ -1,97 +1,144 @@
...
+%End
     virtual int dimension() const;
     virtual QString geometryType() const;
     virtual void clear();
-    virtual QgsAbstractGeometry* boundary() const /Factory/;
+    virtual QgsAbstractGeometry *boundary() const /Factory/;
 
-    /** Adds a geometry and takes ownership. Returns true in case of success.*/
-    virtual bool addGeometry( QgsAbstractGeometry* g /Transfer/ );
+    virtual bool addGeometry( QgsAbstractGeometry *g ) /Transfer/;
```
Comment (nyalldawson) [inline python/core/geometry/qgsgeometrycollection.sip]: transfer in wrong place

File: python/core/geometry/qgsgeometrycollection.sip [hunk 20]
```diff
@@ -1,97 +1,144 @@
...
+Adds a geometry and takes ownership. Returns true in case of success.
+ :rtype: bool
+%End
 
-    /** Inserts a geometry before a specified index and takes ownership. Returns true in case of success.
-     * @param g geometry to insert. Ownership is transferred to the collection.
-     * @param index position to insert geometry before
-     */
-    virtual bool insertGeometry( QgsAbstractGeometry* g /Transfer/, int index );
+    virtual bool insertGeometry( QgsAbstractGeometry *g, int index ) /Transfer/;
```
Comment (nyalldawson) [inline python/core/geometry/qgsgeometrycollection.sip]: misplaced transfer

File: python/core/geometry/qgsgeometrycollection.sip [hunk 21]
```diff
@@ -1,97 +1,144 @@
...
-    virtual QList< QList< QList< QgsPointV2 > > > coordinateSequence() const;
+    virtual QgsCoordinateSequence coordinateSequence() const;
     virtual int nCoordinates() const;
-    virtual double closestSegment( const QgsPointV2& pt, QgsPointV2& segmentPt,  QgsVertexId& vertexAfter, bool* leftOf, double epsilon ) const;
-    bool nextVertex( QgsVertexId& id, QgsPointV2& vertex ) const;
 
-    //low-level editing
-    virtual bool insertVertex( QgsVertexId position, const QgsPointV2& vertex );
-    virtual bool moveVertex( QgsVertexId position, const QgsPointV2& newPos );
+    virtual double closestSegment( const QgsPointV2 &pt, QgsPointV2 &segmentPt,  QgsVertexId &vertexAfter, bool *leftOf, double epsilon ) const;
```
Comment (nyalldawson) [inline python/core/geometry/qgsgeometrycollection.sip]: Should these have /Out/s?

File: python/core/geometry/qgsgeometryengine.sip [hunk 22]
```diff
@@ -1,67 +1,187 @@
...
 class QgsGeometryEngine
 {
-%TypeHeaderCode
-#include <qgsgeometryengine.h>
+%Docstring
+ Contains geometry relation and modification algorithms.
+.. versionadded:: 2.10
+.. note::
+
+   this API is not considered stable and may change for 2.12
```
Comment (nyalldawson) [inline python/core/geometry/qgsgeometryengine.sip]: remove

File: python/core/geometry/qgsgeometryengine.sip [hunk 23]
```diff
@@ -1,67 +1,187 @@
...
-
-    virtual double area( QString* errorMsg = 0 ) const = 0;
-    virtual double length( QString* errorMsg = 0 ) const = 0;
-    virtual bool isValid( QString* errorMsg = 0 ) const = 0;
-    virtual bool isEqual( const QgsAbstractGeometry& geom, QString* errorMsg = 0 ) const = 0;
-    virtual bool isEmpty( QString* errorMsg = 0 ) const = 0;
-
-    virtual int splitGeometry( const QgsLineString& splitLine,
-                               QList<QgsAbstractGeometry*>& newGeometries,
+    virtual QgsAbstractGeometry *intersection( const QgsAbstractGeometry &geom, QString *errorMsg = 0 ) const = 0;
```
Comment (nyalldawson) [inline python/core/geometry/qgsgeometryengine.sip]: I suspect ALL these methods are missing /Factory/s

File: python/core/geometry/qgsgeometryutils.sip [hunk 24]
```diff
@@ -1,82 +1,355 @@
...
+.. versionadded:: 2.10
 %End
 
+%TypeHeaderCode
+#include "qgsgeometryutils.h"
+%End
   public:
 
-    static QList<QgsLineString*> extractLineStrings( const QgsAbstractGeometry* geom );
+    static QList<QgsLineString *> extractLineStrings( const QgsAbstractGeometry *geom );
```
Comment (nyalldawson) [inline python/core/geometry/qgsgeometryutils.sip]: Factory

File: python/core/geometry/qgslinestring.sip [hunk 25]
```diff
@@ -1,156 +1,243 @@
...
+ \param tolerance segmentation tolerance
+ \param toleranceType maximum segmentation angle or maximum difference between approximation and curve*
+ :rtype: QgsLineString
+%End
+
+    virtual int numPoints() const;
+
     virtual int nCoordinates() const;
-    void points( QList<QgsPointV2>& pt ) const;
+    virtual void points( QgsPointSequence &pt ) const;
```
Comment (nyalldawson) [inline python/core/geometry/qgslinestring.sip]: should be /Out/ I think

File: python/core/geometry/qgslinestring.sip [hunk 26]
```diff
@@ -1,156 +1,243 @@
...
-    virtual bool insertVertex( QgsVertexId position, const QgsPointV2& vertex );
-    virtual bool moveVertex( QgsVertexId position, const QgsPointV2& newPos );
+    virtual bool insertVertex( QgsVertexId position, const QgsPointV2 &vertex );
+    virtual bool moveVertex( QgsVertexId position, const QgsPointV2 &newPos );
     virtual bool deleteVertex( QgsVertexId position );
 
-    virtual QgsLineString* reversed() const /Factory/;
+    virtual QgsLineString *reversed() const /Factory/;
+
+    virtual double closestSegment( const QgsPointV2 &pt, QgsPointV2 &segmentPt,  QgsVertexId &vertexAfter, bool *leftOf, double epsilon ) const;
```
Comment (nyalldawson) [inline python/core/geometry/qgslinestring.sip]: missing some Outs?

File: python/core/geometry/qgsmulticurve.sip [hunk 27]
```diff
@@ -1,31 +1,67 @@
...
+class QgsMultiCurve: QgsGeometryCollection
 {
-%TypeHeaderCode
-#include <qgsmulticurve.h>
+%Docstring
+ Multi curve geometry collection.
+.. versionadded:: 2.10
+.. note::
+
+   this API is not considered stable and may change for 2.12
```
Comment (nyalldawson) [inline python/core/geometry/qgsmulticurve.sip]: remove

File: python/core/geometry/qgsmulticurve.sip [hunk 28]
```diff
@@ -1,31 +1,67 @@
...
+%Docstring
+inherited: QString asWkt( int precision = 17 ) const;
+ :rtype: QDomElement
+%End
+    virtual QDomElement asGML3( QDomDocument &doc, int precision = 17, const QString &ns = "gml" ) const;
+
+    virtual QString asJSON( int precision = 17 ) const;
+
+
+    virtual bool addGeometry( QgsAbstractGeometry *g ) /Transfer/;
```
Comment (nyalldawson) [inline python/core/geometry/qgsmulticurve.sip]: transfer in wrong place

File: python/core/geometry/qgsmultilinestring.sip [hunk 29]
```diff
@@ -1,30 +1,69 @@
...
+class QgsMultiLineString: QgsMultiCurve
 {
-%TypeHeaderCode
-#include <qgsmultilinestring.h>
+%Docstring
+ Multi line string geometry collection.
+.. versionadded:: 2.10
+.. note::
+
+   this API is not considered stable and may change for 2.12
```
Comment (nyalldawson) [inline python/core/geometry/qgsmultilinestring.sip]: remove

File: python/core/geometry/qgsmultilinestring.sip [hunk 30]
```diff
@@ -1,30 +1,69 @@
...
+%Docstring
+inherited: QString asWkt( int precision = 17 ) const;
+ :rtype: QDomElement
+%End
+    virtual QDomElement asGML3( QDomDocument &doc, int precision = 17, const QString &ns = "gml" ) const;
+
+    virtual QString asJSON( int precision = 17 ) const;
+
+
+    virtual bool addGeometry( QgsAbstractGeometry *g ) /Transfer/;
```
Comment (nyalldawson) [inline python/core/geometry/qgsmultilinestring.sip]: transfer placement

File: python/core/geometry/qgsmultipoint.sip [hunk 31]
```diff
@@ -1,29 +1,66 @@
...
+
+
+class QgsMultiPointV2: QgsGeometryCollection
 {
+%Docstring
+ Multi point geometry collection.
+.. versionadded:: 2.10
+.. note::
+
+   this API is not considered stable and may change for 2.12
```
Comment (nyalldawson) [inline python/core/geometry/qgsmultipoint.sip]: remove

File: python/core/geometry/qgsmultipoint.sip [hunk 32]
```diff
@@ -1,29 +1,66 @@
...
+    virtual QDomElement asGML3( QDomDocument &doc, int precision = 17, const QString &ns = "gml" ) const;
+
+    virtual QString asJSON( int precision = 17 ) const;
+
 
     virtual int nCoordinates() const;
 
-    /** Adds a geometry and takes ownership. Returns true in case of success*/
-    virtual bool addGeometry( QgsAbstractGeometry* g /Transfer/ );
+    virtual bool addGeometry( QgsAbstractGeometry *g ) /Transfer/;
```
Comment (nyalldawson) [inline python/core/geometry/qgsmultipoint.sip]: transfer placement

File: python/core/geometry/qgsmultipolygon.sip [hunk 33]
```diff
@@ -1,32 +1,71 @@
...
+
+
+class QgsMultiPolygonV2: QgsMultiSurface
 {
+%Docstring
+ Multi polygon geometry collection.
+.. versionadded:: 2.10
+.. note::
+
+   this API is not considered stable and may change for 2.12
```
Comment (nyalldawson) [inline python/core/geometry/qgsmultipolygon.sip]: remove

File: python/core/geometry/qgsmultipolygon.sip [hunk 34]
```diff
@@ -1,32 +1,71 @@
...
+    virtual QDomElement asGML3( QDomDocument &doc, int precision = 17, const QString &ns = "gml" ) const;
 
-    /** Returns the geometry converted to the more generic curve type QgsMultiSurface
-    @return the converted geometry. Caller takes ownership*/
-    QgsAbstractGeometry* toCurveType() const /Factory/;
+    virtual QString asJSON( int precision = 17 ) const;
 
-    virtual QgsAbstractGeometry* boundary() const /Factory/;
+
+    virtual bool addGeometry( QgsAbstractGeometry *g ) /Factory/;
```
Comment (nyalldawson) [inline python/core/geometry/qgsmultipolygon.sip]: should be /Transfer/

File: python/core/geometry/qgsmultisurface.sip [hunk 35]
```diff
@@ -1,27 +1,60 @@
...
+
+
+class QgsMultiSurface: QgsGeometryCollection
 {
+%Docstring
+ Multi surface geometry collection.
+.. versionadded:: 2.10
+.. note::
+
+   this API is not considered stable and may change for 2.12
```
Comment (nyalldawson) [inline python/core/geometry/qgsmultisurface.sip]: remove

File: python/core/geometry/qgsmultisurface.sip [hunk 36]
```diff
@@ -1,27 +1,60 @@
...
+%End
+    virtual QDomElement asGML3( QDomDocument &doc, int precision = 17, const QString &ns = "gml" ) const;
 
-    /** Returns a geometry without curves. Caller takes ownership*/
-    QgsAbstractGeometry* segmentize() const /Factory/;
+    virtual QString asJSON( int precision = 17 ) const;
+
+
+
+    virtual bool addGeometry( QgsAbstractGeometry *g )  /Transfer/;
```
Comment (nyalldawson) [inline python/core/geometry/qgsmultisurface.sip]: transfer placement

File: python/core/geometry/qgspointv2.sip [hunk 37]
```diff
@@ -1,323 +1,365 @@
...
-%End
-
-    bool operator==( const QgsPointV2& pt ) const;
-    bool operator!=( const QgsPointV2& pt ) const;
-
-    /** Returns the point's x-coordinate.
-     * @see setX()
-     * @see rx()
-     */
+%Docstring
```
Comment (nyalldawson) [inline python/core/geometry/qgspointv2.sip]: Missing MethodCode

File: python/core/geometry/qgspointv2.sip [hunk 38]
```diff
@@ -1,323 +1,365 @@
...
+low-level editing
+ :rtype: bool
+%End
+    virtual bool moveVertex( QgsVertexId position, const QgsPointV2 &newPos );
     virtual bool deleteVertex( QgsVertexId position );
 
-    double closestSegment( const QgsPointV2& pt, QgsPointV2& segmentPt,  QgsVertexId& vertexAfter, bool* leftOf, double epsilon ) const;
-    bool nextVertex( QgsVertexId& id, QgsPointV2& vertex ) const;
+    virtual double closestSegment( const QgsPointV2 &pt, QgsPointV2 &segmentPt,  QgsVertexId &vertexAfter, bool *leftOf, double epsilon ) const;
```
Comment (nyalldawson) [inline python/core/geometry/qgspointv2.sip]: Outs?

File: python/core/geometry/qgspointv2.sip [hunk 39]
```diff
@@ -1,323 +1,365 @@
...
+
+    virtual bool nextVertex( QgsVertexId &id, QgsPointV2 &vertex ) const;
```
Comment (nyalldawson) [inline python/core/geometry/qgspointv2.sip]: /Out/ here?

File: python/core/geometry/qgspolygon.sip [hunk 40]
```diff
@@ -1,46 +1,89 @@
...
-    /** Returns the geometry converted to the more generic curve type QgsCurvePolygon
-     @return the converted geometry. Caller takes ownership*/
-    QgsAbstractGeometry* toCurveType() const /Factory/;
+%Docstring
+ Returns the geometry converted to the more generic curve type QgsCurvePolygon
+:return: the converted geometry. Caller takes ownership*
+ :rtype: QgsAbstractGeometry
+%End
+
+    virtual void addInteriorRing( QgsCurve *ring ) /Transfer/;
```
Comment (nyalldawson) [inline python/core/geometry/qgspolygon.sip]: Transfer placement

File: python/core/geometry/qgspolygon.sip [hunk 41]
```diff
@@ -1,46 +1,89 @@
...
 
-    void addInteriorRing( QgsCurve* ring /Transfer/ );
-    //overridden to handle LineString25D rings
-    virtual void setExteriorRing( QgsCurve* ring /Transfer/ );
+    virtual void setExteriorRing( QgsCurve *ring ) /Transfer/;
```
Comment (nyalldawson) [inline python/core/geometry/qgspolygon.sip]: Transfer placement

File: python/core/geometry/qgstriangle.sip [hunk 42]
```diff
@@ -1,63 +1,334 @@
...
+
+%Docstring
+inherited: QString asJSON( int precision = 17 ) const;
+ :rtype: QgsPolygonV2
+%End
+
+    virtual QgsAbstractGeometry *toCurveType() const /Factory/;
+
+
+    virtual void addInteriorRing( QgsCurve *ring ) /Transfer/;
```
Comment (nyalldawson) [inline python/core/geometry/qgstriangle.sip]: Transfer placement

File: python/core/geometry/qgstriangle.sip [hunk 43]
```diff
@@ -1,63 +1,334 @@
...
+ :rtype: bool
+%End
+    virtual bool moveVertex( QgsVertexId vId, const QgsPointV2 &newPos );
+
 
-    void addInteriorRing( QgsCurve* ring /Transfer/ ); // NOTE: no interior ring for triangle.
-    bool deleteVertex( QgsVertexId position );
-    bool insertVertex( QgsVertexId position, const QgsPointV2 &vertex );
-    bool moveVertex( QgsVertexId vId, const QgsPointV2& newPos );
+    virtual void setExteriorRing( QgsCurve *ring ) /Transfer/;
```
Comment (nyalldawson) [inline python/core/geometry/qgstriangle.sip]: Transfer placement

**Comment by nyalldawson** [comment]
I realise not all these comments are new issues, but they should be addressed too :)

File: python/core/geometry/qgsabstractgeometry.sip (hunk 1 above)
Comment (3nids) [inline python/core/geometry/qgsabstractgeometry.sip]: this is commented in the source too
https://github.com/qgis/QGIS/blob/7c79803cc37bd2bd79ca5c0c2d7f61b46eab61bc/python/core/geometry/qgsabstractgeometry.sip#L105

File: python/core/geometry/qgsgeometry.sip (hunk 12 above)
Comment (3nids) [inline python/core/geometry/qgsgeometry.sip]: issue in the header itself

**Comment by 3nids** [comment]
I've found a few more Outs...

**Comment by 3nids** [comment]
almost ready to merge.
@nyalldawson sip doc coverage is failing:
> Missing members:
  QgsConstWkbPtr.read
  QgsConstWkbPtr.readHeader
  QgsConstWkbPtr.remaining
  QgsConstWkbPtr.verifyBound
  QgsWkbPtr.remaining
  QgsWkbPtr.size
  QgsWkbPtr.writtenSize

What's the correct way to fix these, since they don't have a docstring at the moment?

**Comment by nyalldawson** [comment]
> What's the correct way to fix these, since they don't have a docstring at the moment?

Add a docstring ;)

**Pull request merged**